aten/src/ATen/native/Normalization.cpp

```diff
@@ -615,4 +635,39 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_backward_cpu(const Tensor& grad_ou
     });
 }
 
+TORCH_IMPL_FUNC(renorm_out)(const Tensor& self, const Scalar& p, int64_t dim,
+                            const Scalar& maxnorm, const Tensor& out) {
+  auto self_sizes = self.sizes();
+  dim = c10::maybe_wrap_dim(dim, self_sizes.size());
+
+  DimVector reduce_dims(self_sizes.size());
+  std::iota(reduce_dims.begin(), reduce_dims.end(), 0);
+  reduce_dims.erase(reduce_dims.begin() + dim);
+
+  // For cuda half, calculate norm in float precision then cast
+  // normalization factor to half
+  auto dtype = self.scalar_type();
+  auto acc_type = at::toAccumulateType(dtype, /*is_cuda=*/true);
+  Tensor norm;
+  if (acc_type != dtype) {
+    norm = at::linalg_vector_norm(self, p.toDouble(), reduce_dims,
+                                  /*keepdim=*/true, /*dtype=*/acc_type);
+  } else {
+    norm = at::linalg_vector_norm(self, p.toDouble(), reduce_dims,
+                                  /*keepdim=*/true);
+  }
```

Comment:
For complex types, the input is cast to `dtype` and the output has type `dtype` so it has to be complex. However, if I leave it out entirely, the result is real. Seems like an odd way for this to work.

Reply:
cc @kurtamohler for norm `dtype` behavior. Should we disallow specifying non-complex `dtype` for complex? I believe there was a warning about that at some point. But then, it doesn't allow users to specify different precision if the input is complex.

Reply:
I agree that it's odd behavior. The explanation is that the input is converted to `dtype` before the norm is calculated, so if we allowed a non-complex `dtype` arg when the input is complex, we would be discarding the imaginary part. So at the moment we throw an error in that case: https://github.com/pytorch/pytorch/blob/afdfd2288ab/aten/src/ATen/native/LinearAlgebra.cpp#L2271

Reply:
If the result was always real then `norm(tensor, ..., dtype=dtype)` would behaave just like `norm(tensor.to(dtype), ...)`. I think that's reasonable.